### PR TITLE
Add apisupport with two checks

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
Depends on #446.

**Comment by gperciva** [comment]
In #83, you mused about adding support for other APIs (in particular, `getrandom(3)`).  `apisupport.sh` sets of the framework which would allow for that.

At the moment, this PR only uses it for detecting some nitpicky compiler things.  In a few places, we don't strictly obey the C99 & POSIX standards, so some compilers legitimately complain about it.

This PR does not add the capability for our code to do things like `#ifdef APISUPPORT_NONPOSIX_SETGROUPS`.  I have that available in a separate branch, but at the moment there's no need for such things, so I haven't included it.  Likewise, there's no runtime checks (such as a theoretical `apisupport_linux_getrandom()`, similar to `cpusupport_x86_shani()`) in this particular PR.

File: apisupport/Build/apisupport.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+fi
+# Standard output should be written to apisupport-config.h, which is both a
+# C header file defining APISUPPORT_ARCH_FEATURE macros and sourceable sh
+# code which sets CFLAGS_ARCH_FEATURE environment variables.
+SRCDIR=$(command -p dirname "$0")
+
+CFLAGS_HARDCODED="-D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=700"
+
+feature() {
+	ARCH=$1
```
Comment (cperciva) [inline apisupport/Build/apisupport.sh:15]: s/ARCH/PLATFORM/ ?

Comment (gperciva) [reply apisupport/Build/apisupport.sh:15]: Hmm, good point.  I'm happy calling it PLATFORM, but at the moment, I'm using that variable to be `NONPOSIX` and `LIBSSL`.  Is that stretching the meaning of "platform"?

Comment (cperciva) [reply apisupport/Build/apisupport.sh:15]: Maybe a bit of a stretch but it's better than anything else which comes to mind.

Comment (gperciva) [reply apisupport/Build/apisupport.sh]: ok, done as a REBASE commit.

**Comment by cperciva** [comment]
Ok, LGTM.  We should probably change the NO_SETGROUPS check to a APISUPPORT_NONPOSIX_SETGROUPS in setgroups_none.c at some point though?

**Comment by gperciva** [comment]
At some point, yes.

> This PR does not add the capability for our code to do things like `#ifdef APISUPPORT_NONPOSIX_SETGROUPS`

Although maybe "some point" should be "now".

It would be easy to add that back in -- actually, the first 3 drafts of this PR contained that code, but I removed it since it wasn't needed.  This gives me an excuse to get that infrastructure done now.

**Comment by gperciva** [comment]
> We should probably change the NO_SETGROUPS check to a APISUPPORT_NONPOSIX_SETGROUPS in setgroups_none.c at some point though?

On second thought, I'm not sure about this.

Suppose that FreeBSD 14.0 moved `setgroups()` out of `<unistd.h>` and into `<grp.h>`.  What would we want libcperciva to do?

* give a compile error (because we expected `util/setgroups_none.c` to have the correct includes)
* silently make `setgroups_none()` do nothing

I'd say that "compile error" is the appropriate response.

I'll still add the "support preprocessor `#if APISUPPORT_FOO_BAR`" stuff to this PR, but I won't mess with NO_SETGROUPS.  We can revisit that in a separate PR targeted at that case.

**Comment by gperciva** [comment]
Updated.  I put the `cpusupport.sh` changes in a separate #449, but if you'd prefer to just merge it all in this PR, that's fine with me.

There's now a ton of `Makefile` changes to add `-DAPISUPPORT_CONFIG_FILE=\"apisupport-config.h\"` everywhere.

File: apisupport/Build/apisupport.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+SRCDIR=$(command -p dirname "$0")
+
+CFLAGS_HARDCODED="-D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=700"
+
+feature() {
+	PLATFORM=$1
+	FEATURE=$2
+	EXTRALIB=$3
+	shift 3;
+	if ! [ -f "${SRCDIR}"/apisupport-"$PLATFORM"-"$FEATURE".c ]; then
```
Comment (cperciva) [inline apisupport/Build/apisupport.sh:19]: Excessive quotes (as in cpusupport.sh).

**Comment by gperciva** [comment]
Fixed, rebased, ready for merge or more review.

**Pull request merged**